File: back/src/routes/admin/stock/product_management.rs
use axum::{
    extract::State,
    routing::{get, patch, post},
    Json, Router,
};
use serde::Deserialize;

use crate::{
    admin::user::AdminUser,
    app::products::{self, MoveDirection, Product},
    errors::{ManageStockError, ServerError},
    routes::{extractors::CustomQuery as Query, reponders::OkEmptyResponse, AppState},
    utils::deserialize_empty_as_none,
};

pub fn get_router() -> Router<AppState> {
    Router::new()
        .route(
            "/",
            post(insert_product)
                .patch(edit_product)
                .delete(delete_product),
        )
        .route("/get_all", get(get_all_products))
        .route("/move", patch(move_product))
        .route("/add_variation", post(add_variation))
        .route("/remove_variation", post(remove_variation))
}

#[derive(Deserialize)]
struct InsertProductParams {
    name: String,
    description: String,
    stock_quantity: f32,
}
async fn insert_product(
    State(state): State<AppState>,
    _user: AdminUser,
    params: Query<InsertProductParams>,
) -> Result<OkEmptyResponse, ManageStockError> {
    products::Product::create(
        &state.pool,
        params.name.clone(),
        params.description.clone(),
        params.stock_quantity,
    )
    .await?;

    Ok(OkEmptyResponse::new())
}

#[derive(Deserialize)]
struct EditProductParams {
    product_id: u32,
    #[serde(default, deserialize_with = "deserialize_empty_as_none")]
    new_name: Option<String>,
    #[serde(default, deserialize_with = "deserialize_empty_as_none")]
    new_description: Option<String>,
    #[serde(default, deserialize_with = "deserialize_empty_as_none")]
    new_stock_quantity: Option<f32>,
}

async fn edit_product(
    State(state): State<AppState>,
    _user: AdminUser,
    params: Query<EditProductParams>,
) -> Result<OkEmptyResponse, ManageStockError> {
    let mut product = match Product::get(&state.pool, params.product_id).await? {
        Some(p) => p,
        None => return Err(ManageStockError::ProductNotFound(params.product_id)),
    };

    if let Some(new_name) = &params.new_name {
        product.set_name(&state.pool, new_name.to_owned()).await?;
    }
    if let Some(new_description) = &params.new_description {
        product
            .set_description(&state.pool, new_description.to_owned())
            .await?;
    }
    if let Some(new_stock_quantity) = params.new_stock_quantity {
        product
            .set_stock_quantity(&state.pool, new_stock_quantity)
            .await?;
    }

    Ok(OkEmptyResponse::new())
}

#[derive(Deserialize)]
struct DeleteProductParams {
    product_id: u32,
}
async fn delete_product(
    State(state): State<AppState>,
    _user: AdminUser,
    params: Query<DeleteProductParams>,
) -> Result<OkEmptyResponse, ManageStockError> {
    let product = match Product::get(&state.pool, params.product_id).await? {
        Some(p) => p,
        None => return Err(ManageStockError::ProductNotFound(params.product_id)),
    };
    product.delete(&state.pool).await?;

    Ok(OkEmptyResponse::new())
}
async fn get_all_products(
    State(state): State<AppState>,
) -> Result<Json<Vec<products::Product>>, ServerError> {
    let products = products::get_all(&state.pool).await?;
    Ok(Json(products))
}

#[derive(Deserialize)]
struct MoveProductParams {
    product_id: u32,
    direction: MoveDirection,
}
async fn move_product(
    State(state): State<AppState>,
    _user: AdminUser,
    params: Query<MoveProductParams>,
) -> Result<OkEmptyResponse, ManageStockError> {
    let mut product = match Product::get(&state.pool, params.product_id).await? {
        Some(p) => p,
        None => return Err(ManageStockError::ProductNotFound(params.product_id)),
    };
    product.move_product(&state.pool, params.direction).await?;

    Ok(OkEmptyResponse::new())
}

#[derive(Deserialize)]
struct AddVariationParams {
    product_id: u32,
    name: String,
    price_ht: i32,
    tva: f32,
    volume: f32,
    available_to_order: bool,
}
async fn add_variation(
    State(state): State<AppState>,
    _user: AdminUser,
    params: Query<AddVariationParams>,
) -> Result<OkEmptyResponse, ManageStockError> {
    let mut product = match Product::get(&state.pool, params.product_id).await? {
        Some(p) => p,
        None => return Err(ManageStockError::ProductNotFound(params.product_id)),
    };

    product
        .add_variation(
            &state.pool,
            params.name.to_owned(),
            params.price_ht,
            params.tva,
            params.volume,
            params.available_to_order,
        )
        .await?;

    Ok(OkEmptyResponse::new())
}

#[derive(Deserialize)]
struct RemoveVariationParams {
    product_id: u32,
    variation_id: u32,
}
async fn remove_variation(
    State(state): State<AppState>,
    _user: AdminUser,
    params: Query<RemoveVariationParams>,
) -> Result<OkEmptyResponse, ManageStockError> {
    let mut product = match Product::get(&state.pool, params.product_id).await? {
        Some(p) => p,
        None => return Err(ManageStockError::ProductNotFound(params.product_id)),
    };
    product
        .delete_variation(&state.pool, params.variation_id)
        .await?;

    Ok(OkEmptyResponse::new())
}

File: front/src/scripts/api/admin/stock/product-management.ts
import { base, Error } from "../../api";
import type { Product } from "../../products";

export async function get_all_products(): Promise<Product[]> {
  let url = `${base}/admin/stock/products/get_all`;
  let error_title = "Erreur lors de la recupération du stock";
  try {
    let res = await fetch(url).then(async e => await e.json())
    if (res.error) {
      new Error(error_title, res.error)
      return []
    } else {
      let prods = res as Product[];
      for(let i=0; i<prods.length; i++){
        for(let j=0; j<prods[i].variations.length; j++){
          prods[i].variations[j].price_ttc =  prods[i].variations[j].price_ht * (1 + prods[i].variations[j].tva)
        }
      }
      return prods
    }
  } catch (e: any) {
    new Error(error_title, e.toString());
    return []
  }
}
export async function insert_product(new_prod: Product): Promise<boolean> {
    let url = `${base}/admin/stock/products` +
        `?name=${encodeURIComponent(new_prod.name)}` +
        `&description=${encodeURIComponent(new_prod.description)}` +
        `&stock_quantity=${encodeURIComponent(new_prod.stock_quantity)}`;
    let error_title = "Erreur lors de la création d'un nouveau produit"
    try {
        let res = await fetch(url, {
            method: "POST",
            credentials: "include",
        }).then(e => e.json());
        if (res.error) {
            new Error(error_title, res.error)
            return false
        } else {
            return true
        }
    } catch (e: any) {
        new Error(error_title, e.toString());
        return false
    }
}

export type ProductEdition = {
    new_name?: string,
    new_description?: string,
    new_stock_quantity?: number,
}
export async function edit_product(
    product_id: number,
    edition: ProductEdition
): Promise<boolean> {
    let url = `${base}/admin/stock/products?product_id=${encodeURIComponent(product_id)}` +
        `&new_name=${encodeURIComponent(edition.new_name ? edition.new_name : "")}` +
        `&new_description=${encodeURIComponent(edition.new_description ? edition.new_description : "")}` +
        `&new_stock_quantity=${encodeURIComponent(edition.new_stock_quantity ? edition.new_stock_quantity : "")}`;
    let error_title = "Erreur lors de la maj du stock"
    try {
        let res = await fetch(url, {
            method: "PATCH",
            credentials: "include",
        }).then(e => e.json());
        if (res.error) {
            new Error(error_title, res.error)
            return false
        } else {
            return true
        }
    } catch (e: any) {
        new Error(error_title, e.toString());
        return false
    }
}


export async function delete_product(product_id: number): Promise<boolean> {
    let url = `${base}/admin/stock/products?product_id=${encodeURIComponent(product_id)}`;
    let error_title = "Erreur lors de la suppréssion d'un produit"
    try {
        let res = await fetch(url, {
            method: "DELETE",
            credentials: "include",
        }).then(e => e.json());
        if (res.error) {
            new Error(error_title, res.error)
            return false
        } else {
            return true
        }
    } catch (e: any) {
        new Error(error_title, e.toString());
        return false
    }
}

export async function move_product(product_id: number, direction: "up" | "down"): Promise<boolean> {
    let url = `${base}/admin/stock/products/move?product_id=${encodeURIComponent(product_id)}&direction=${encodeURIComponent(direction)}`;
    let error_title = "Erreur lors du déplacement d'un produit dans la liste"
    try {
        let res = await fetch(url, {
            method: "PATCH",
            credentials: "include",
        }).then(e => e.json());
        if (res.error) {
            new Error(error_title, res.error)
            return false
        } else {
            return true
        }
    } catch (e: any) {
        new Error(error_title, e.toString());
        return false
    }
}


export async function add_variation(product_id: number, name: string, price_ht: number, tva: number, volume: number, available_to_order: boolean): Promise<boolean> {
    let url = `${base}/admin/stock/products/add_variation?product_id=${encodeURIComponent(product_id)}` +
        `&name=${encodeURIComponent(name)}` +
        `&price_ht=${encodeURIComponent(price_ht)}` +
        `&tva=${encodeURIComponent(tva)}` +
        `&available_to_order=${encodeURIComponent(available_to_order)}` +
        `&volume=${encodeURIComponent(volume)}`;
    let error_title = "Erreur lors de l'ajout d'une variation"
    try {
        let res = await fetch(url, {
            method: "POST",
            credentials: "include",
        }).then(e => e.json());
        if (res.error) {
            new Error(error_title, res.error)
            return false
        } else {
            return true
        }
    } catch (e: any) {
        new Error(error_title, e.toString());
        return false
    }
}

export async function remove_variation(product_id: number, variation_id: number): Promise<boolean> {
    let url = `${base}/admin/stock/products/remove_variation?product_id=${encodeURIComponent(product_id)}` +
        `&variation_id=${encodeURIComponent(variation_id)}`;
    let error_title = "Erreur lors du retrait d'une variation"
    try {
        let res = await fetch(url, {
            method: "POST",
            credentials: "include",
        }).then(e => e.json());
        if (res.error) {
            new Error(error_title, res.error)
            return false
        } else {
            return true
        }
    } catch (e: any) {
        new Error(error_title, e.toString());
        return false
    }
}
